pyinv/assets/models/asset.py

```diff
@@ -41,3 +44,30 @@ def asset_codes(self) -> List[str]:
 
     def __str__(self) -> str:
         return f"{self.display_name} ({self.first_asset_code})"
+
+    def add_asset_code(self, code_type: AssetCodeType, code: Optional[str]) -> AssetCode:
+        """
+        Add an asset code to an asset.
+
+        :raises ValueError: Unable to generate code, or unrecognised code type.
+        :raises django.db.IntegrityError: The specifed code already exists
+        """
+        def _generate_code(strategy: AssetCodeStrategy) -> Optional[str]:
+            code = strategy.generate_new_code()
+
+            # If we have generated a code, check it's not already in use.
+            if code is not None and AssetCode.objects.filter(code=code).exists():
+                # Try again.
+                return _generate_code(strategy)  # pragma: nocover
+
+            return code
+
+        if code is None:
+            code = _generate_code(AssetCodeType.get_strategy(code_type))
+
+            if code is None:
+                raise ValueError(
+                    "Unable to generate an asset code of that type. Please specify your own.",
+                )
+
+        return AssetCode.objects.create(asset=self, code=code, code_type=code_type.value)
```

Comment:
Issue: This code isn't fully thread safe. There's a period of time in between checking whether a code exists and actually trying to save where a code may be inserted. You're best bet is to either repeatedly call `.create` rather than checking, or to use `select_for_update` to lock the table whilst you generate an id (not a good idea)

Reply:
[6ab2efa](https://github.com/pyinv/pyinv-server/pull/10/commits/6ab2efa8595e3c0727170098f932b4fad769329b)